evm_loader/program/src/types/transaction.rs

```diff
@@ -117,6 +135,299 @@ impl rlp::Decodable for Transaction {
+            if entry.is_list() {
+                // Parse address from first element
+                let address: Address = entry.at(0)?.as_val()?;
+
+                // Get storage keys from second element
+                let mut storage_keys: Vec<StorageKey> = vec![];
+
+                for key in entry.at(1)?.iter() {
+                    storage_keys.push(key.as_val()?);
+                }
+
+                access_list.push((address, storage_keys));
+            }
+        }
+
+        let y_parity: u8 = rlp.at(8)?.as_val()?;
+        let r: U256 = u256(&rlp.at(9)?)?;
+        let s: U256 = u256(&rlp.at(10)?)?;
+
+        if rlp.at(11).is_ok() {
+            return Err(rlp::DecoderError::RlpIncorrectListLen);
+        }
+
+        let hash = solana_program::keccak::hashv(&[&[0x01], rlp.as_raw()]).to_bytes();
+        let signed_hash = eip2718_signed_hash(&[0x01], rlp, 8)?;
+
+        let tx = AccessListTx {
+            nonce,
+            gas_price,
+            gas_limit,
+            target,
+            value,
+            call_data,
+            r,
+            s,
+            chain_id,
+            recovery_id: y_parity,
+            access_list,
+            rlp_len,
+            hash,
+            signed_hash,
+        };
+
+        Ok(tx)
+    }
+}
+
+// TODO: Will be added as a part of EIP-1559
+// struct DynamicFeeTx {}
+
+// TODO: Will be added as a part of EIP-1559
+// struct BlobTx {}
+
+#[derive(Debug, Clone)]
+pub enum Transaction {
+    Legacy(LegacyTx),
+    AccessList(AccessListTx),
+}
+
+impl Transaction {
+    pub fn from_rlp(transaction: &[u8]) -> Result<Self, Error> {
+        let (transaction_type, transaction) = TransactionEnvelope::get_type(transaction);
+
+        let tx = match transaction_type {
+            TransactionEnvelope::Legacy => {
+                Transaction::Legacy(rlp::decode::<LegacyTx>(transaction).map_err(Error::from)?)
+            }
+            TransactionEnvelope::AccessList => Transaction::AccessList(
+                rlp::decode::<AccessListTx>(transaction).map_err(Error::from)?,
+            ),
+            _ => unimplemented!(),
+        };
+
+        Ok(tx)
+    }
+
+    pub fn recover_caller_address(&self) -> Result<Address, Error> {
+        use solana_program::keccak::{hash, Hash};
+        use solana_program::secp256k1_recover::secp256k1_recover;
+
+        let signature = [self.r().to_be_bytes(), self.s().to_be_bytes()].concat();
+        let public_key = secp256k1_recover(self.signed_hash(), self.recovery_id(), &signature)?;
+
+        let Hash(address) = hash(&public_key.to_bytes());
+        let address: [u8; 20] = address[12..32].try_into()?;
+
+        Ok(Address::from(address))
+    }
+
+    #[must_use]
+    pub fn nonce(&self) -> u64 {
+        match self {
+            Transaction::Legacy(LegacyTx { nonce, .. })
+            | Transaction::AccessList(AccessListTx { nonce, .. }) => *nonce,
+        }
+    }
+
+    #[must_use]
+    pub fn gas_price(&self) -> &U256 {
```

Comment:
`U256` is trivial copyable type. It's better to return it by value.
Returning it by reference will only cause you to use dereference everywhere.

Reply:
done